Add PubSub lib for Prow to report

yutongz opened the pull request:
We want to extend Prow's ability to hook up with different system. Using [Cloud Pub/Sub](https://cloud.google.com/pubsub/) is one of the steps.

The Report() takes in a finished prowjob, and if is embedded with Pub/Sub related labels (pubsub-project, pubsub-topic and unique run-id), it generates pub/sub messages based on this prowjob and publish to the target Pub/Sub topic.

This lib will be added to crier in a separate PR.

yutongz commented:
/cc @krzyzacy

yutongz commented:
The first commit is actual logic. 
I have problem with deps update, which is in second commit. I ran `hack/update-deps.sh --add cloud.google.com/go/pubsub`. I succeeded which bunch of vendor files updated and added.

Then when I tried to `bazel build prow/pubsub/... `, there are some deps failures:
```bash
➜  test-infra git:(pubsub-lib) bazel build prow/pubsub/...                         
INFO: Build options have changed, discarding analysis cache.
INFO: Analysed 4 targets (153 packages loaded).
INFO: Found 4 targets...
ERROR: /usr/local/google/home/yutongz/go/src/k8s.io/test-infra/vendor/google.golang.org/grpc/internal/channelz/BUILD.bazel:3:1: GoCompile vendor/google.golang.org/grpc/internal/channelz/linux_amd64_stripped/go_default_library~/vendor/google.golang.org/grpc/internal/channelz.a failed (Exit 1)
/usr/local/google/home/yutongz/.cache/bazel/_bazel_yutongz/06f9205d9e32279434ab3ddc2635dd8f/sandbox/linux-sandbox/25/execroot/__main__/vendor/google.golang.org/grpc/internal/channelz/types_linux.go:41:15: undefined: unix.GetsockoptLinger
/usr/local/google/home/yutongz/.cache/bazel/_bazel_yutongz/06f9205d9e32279434ab3ddc2635dd8f/sandbox/linux-sandbox/25/execroot/__main__/vendor/google.golang.org/grpc/internal/channelz/types_linux.go:44:15: undefined: unix.GetsockoptTimeval
/usr/local/google/home/yutongz/.cache/bazel/_bazel_yutongz/06f9205d9e32279434ab3ddc2635dd8f/sandbox/linux-sandbox/25/execroot/__main__/vendor/google.golang.org/grpc/internal/channelz/types_linux.go:47:15: undefined: unix.GetsockoptTimeval
/usr/local/google/home/yutongz/.cache/bazel/_bazel_yutongz/06f9205d9e32279434ab3ddc2635dd8f/sandbox/linux-sandbox/25/execroot/__main__/vendor/google.golang.org/grpc/internal/channelz/types_linux.go:50:15: undefined: unix.GetsockoptTCPInfo
GoCompile: error running compiler: exit status 1
INFO: Elapsed time: 4.476s, Critical Path: 0.87s
INFO: 26 processes: 26 linux-sandbox.
FAILED: Build did NOT complete successfully

```

krzyzacy commented:
/ok-to-test

krzyzacy commented:
xref https://github.com/grpc/grpc-go/issues/2181

probably need to update something else, looking

krzyzacy commented:
hummmmm @BenTheElder @ixdy @cblecker dep experts: it probably should also update https://github.com/kubernetes/test-infra/blob/master/Gopkg.lock#L797-L803 according to the grpc issue but it didn't? need to be done manually?

stevekuznetsov commented:
Is there a design doc for this effort? Are we developing a pluggable reporter API? Will it be exposed through config?

stevekuznetsov commented:
Please share the design document to `kubernetes-sig-testing` mailing list and let's get a review for that first before we try to commit any code.

/hold

ixdy commented:
@krzyzacy yeah, you'll need to explicitly update that dependency, since grpc-go isn't using dep.

yutongz commented:
1. Addressed comments on design doc.
2. Make pubsub reporter a separated controller, as discussed in doc. Following https://github.com/kubernetes/test-infra/pull/9300

PTAL

krzyzacy commented:
we'll probably need to address https://github.com/kubernetes/test-infra/pull/9300#discussion_r216474204

yutongz commented:
Since there are still some crier discussions going on, maybe I can make this PR only for pub/sub reporter lib. We can add it to a crier controller or make itself a separated controller in another PR when everything is settled. 

What's your preference? @stevekuznetsov

krzyzacy commented on prow/pubsub/reporter/reporter.go:
```diff
@@ -0,0 +1,105 @@
+func (c *Client) Report(pj *kube.ProwJob) error {
+	message, err := generateMessageFromPJ(pj)
+	if message == nil {
+		return err
```
hummm, if this is a github/gerrit job, it will return nil here which is same as reported successfully, but we don't want pubsub reporter to alter the prowjob report status. We'll need to return something like a "cannot handle by this reporter" error here I think.

krzyzacy commented on prow/pubsub/reporter/reporter.go:
```diff
@@ -0,0 +1,105 @@
+	if projectName == "" || topicName == "" {
+		return nil, nil
+	}
+	runID := pj.GetLabels()[pubsubRunIDLabel]
```
I thought we want to use the pj ID as the uuid here?

yutongz replied:
We agreed to let users take charge of this uuid and make sure the uniqueness, so still going to take from this label.

krzyzacy replied:
ack

krzyzacy commented on prow/Makefile:
```diff
@@ -296,3 +298,12 @@ gerrit-crier-image: alpine-image
+	$(PUSH) "$(REGISTRY)/$(PROJECT)/pubsub-crier:$(PUBSUB_CRIER_VERSION)"
+	rm cmd/pubsub-crier/pubsub-crier
+
+.PHONY: pubsub-crier-image
```
drop all the deployment changes, and let's put it in the same process in a separate PR?

stevekuznetsov commented:
@yutongz happy to merge this if it's just the dep -- one outstanding comment on the design doc

krzyzacy commented:
/lint

yutongz commented:
/lint

yutongz commented:
/lint

yutongz commented:
/lint

krzyzacy commented:
wanna also add an OWNERS file to prow/pubsub and add yourself? :-)

krzyzacy commented:
lgtm otherwise

krzyzacy commented:
(probably easier just re-vendor everything)

yutongz commented:
Done. Sorry I messed up a little, could you help me remove the extra labels added by the robot?

stevekuznetsov commented:
Oh lord what has happened here

yutongz commented:
Yeah, I messed up something while squashing. 😛

stevekuznetsov commented on prow/pubsub/OWNERS:
```diff
@@ -0,0 +1,4 @@
+approvers:
```
Can you add a `area/prow/pubsub` label here and in `label_sync`? This will help you track overall progress here.

yutongz replied:
Done

yutongz commented:
/test pull-test-infra-gubernator

krzyzacy commented on prow/pubsub/reporter/reporter.go:
```diff
@@ -0,0 +1,109 @@
+func (c *Client) Report(pj *kube.ProwJob) error {
+	message, err := generateMessageFromPJ(pj)
+	if message == nil {
+		return fmt.Errorf("this prowjob cannot handle by this reporter, prowjob name: %s", pj.Name)
```
we probably need to classify a prowjob from failed to report, to should not be handled by this reporter.. but we can do as follow ups.

stevekuznetsov replied:
This should not error out. If there is no work to do, just don't do it.

krzyzacy replied:
Let's just add a function like 

```
func (c *Client) ShouldReport(pj *kube.ProwJob) bool {
  return pj.Labels[pubsubProjectLabel] != "" && pj.Labels[pubsubTopicLabel] != ""
}
```

I'll update the template

stevekuznetsov replied:
Why not just return early and exit with a `nil` error? What's special about this case? Why do we need to know no report occurred?

krzyzacy replied:
it doesn't need to update previous reported status and patch the object to apiserver, if it doesn't need to be reported.

stevekuznetsov replied:
SGTM

krzyzacy replied:
I'll patch this in https://github.com/kubernetes/test-infra/pull/9380

fejta commented:
Will you please put the label change in it's own PR?

Also can you split this into two commits -- one for the vendor change and one for the code.

You can use `git reset HEAD~1` to pull all these files back into a staged state.

yutongz commented:
I am confused, should I put the label change in this PR with a separated commit, or put it in a separated PR?
@fejta  @stevekuznetsov

stevekuznetsov commented:
Defer to @fejta -- let's have three separate commits, one of which (`label_sync`) is in a separate PR

yutongz commented:
Separated commits and put label changes into https://github.com/kubernetes/test-infra/pull/9393
PTAL

krzyzacy commented:
/lgtm

yutongz commented:
PTAL

krzyzacy commented:
@stevekuznetsov I vaguely remember you commented something last night but I cannot find them?

krzyzacy commented:
/lgtm

stevekuznetsov commented:
I managed to comment on a commit on my phone, let me do a final pass

stevekuznetsov commented:
/hold cancel

yutongz commented:
Steve commented on a specific commit. So when I fixed those comments and rebase, the commit is assigned with a new SHA, that's why we can't see those comments... But I did fix them 🙂
@krzyzacy

krzyzacy commented:
rebased & integrated this in https://github.com/kubernetes/test-infra/pull/9380
